cpp/ql/src/semmle/code/cpp/models/implementations/StdPair.qll

```diff
@@ -4,6 +4,37 @@
 
 import semmle.code.cpp.models.interfaces.Taint
 
+/**
+ * An instantiation of `std::pair<T1, T2>`.
+ */
+class StdPairClass extends ClassTemplateInstantiation {
+  StdPairClass() { getTemplate().hasQualifiedName("std", "pair") }
+}
+
+/**
+ * Any of the single-parameter constructors of `std::pair` that takes a reference to an
+ * instantiation of `std::pair`. These constructors allow conversion between pair types when the
+ * underlying element types are convertible.
+ */
+class StdPairCopyishConstructor extends Constructor, TaintFunction {
+  StdPairCopyishConstructor() {
+    this.getDeclaringType() instanceof StdPairClass and
+    this.getNumberOfParameters() = 1 and
+    this.getParameter(0).getUnspecifiedType().(ReferenceType).getBaseType().getUnspecifiedType()
```

Comment:
```suggestion
    this.getParameter(0).getUnspecifiedType().(ReferenceType).getBaseType()
```

There should be no need for a second `getUnspecifiedType` call -- the first one is recursive and goes all the way down.